core/src/test/kotlin/com/exactpro/th2/read/db/app/FullReadDbOracleIntegrationTest.kt

```diff
@@ -0,0 +1,308 @@
+        .pins {
+            publishers {
+                pin(MESSAGE_OUT_PIN) {
+                    attributes("transport-group")
+                }
+            }
+        }
+
+    @Suppress("unused")
+    val cradle = CradleSpec.create()
+        .disableAutoPages()
+        .reuseKeyspace()
+
+    @BeforeAll
+    fun init() {
+        oracle.withDatabaseName("test_data")
+            .start()
+    }
+
+    @AfterAll
+    fun cleanup() {
+        oracle.stop()
+    }
+
+    @BeforeEach
+    fun setup() {
+        execute {
+            dropTable()
+            initTestData()
+        }
+    }
+
+    @ParameterizedTest
+    @CsvSource(
+        "$PULL_BY_DATE_QUERY_ID,$DATE_COLUMN",
+        "$PULL_BY_TIMESTAMP_QUERY_ID,$TIMESTAMP_COLUMN",
+//        "$PULL_BY_TIMESTAMP_WITH_TIME_ZONE_QUERY_ID,$TIMESTAMP_WITH_TIME_ZONE_COLUMN",
+//        "$PULL_BY_TIMESTAMP_WITH_LOCAL_TIME_ZONE_QUERY_ID,$TIMESTAMP_WITH_LOCAL_TIME_ZONE_COLUMN",
+    )
+    @CustomConfigProvider("config")
+    fun `pull by date filed type test`(
+        queryId: String,
+        column: String,
+        @Th2AppFactory factory: CommonFactory,
+        @Th2TestFactory test: CommonFactory,
+        resourceCleaner: CleanupExtension.Registry,
+    ) {
+        val publishedMessages = mutableListOf<Message<*>>()
+        val dbPullRequest = DbPullRequest.newBuilder().apply {
+            sourceIdBuilder.setId(RECORDS_DATA_SOURCE_ID)
+            updateQueryIdBuilder.setId(queryId)
+            initParametersBuilder
+                .putValues(column, "1970-01-01 00:00:00.000")
+            addUseColumns(column)
+            startFromLastReadRow = true
+            pullInterval = Durations.fromMillis(200)
+            resetStateParametersBuilder.setAfterDate(Timestamps.now())
+        }.build()
+
+        val dp = DataProviderGrpcServer().apply { test.grpcRouter.startServer(this).start() }
+        test.transportGroupBatchRouter.subscribe({ _, batch ->
+            batch.groups.asSequence()
+                .flatMap { it.messages.asSequence() }
+                .forEach {
+                    check(it is RawMessage) {
+                        "Unexpected message type ${it::class.java}, id: ${it.id.logId}"
+                    }
+                    dp.set(it.id.toProto(batch.book, batch.sessionGroup), it.body.array(), it.metadata)
+                    publishedMessages.add(it)
+                }
+        }, MESSAGE_OUT_PIN)
+
+        setupApp(factory) { name, resource -> resourceCleaner.add(name, resource) }
+        val readDb = test.grpcRouter.getService(ReadDbService::class.java)
+
+        execute { insertData(listOf(Record(1, Instant.now(), Instant.now(), Instant.now(), Instant.now()))) }
+        var dbPullResponse = readDb.startPulling(dbPullRequest)
+        await("first message").atMost(5, TimeUnit.SECONDS).until { publishedMessages.size == 1 }
+        readDb.stopPulling(StopPullingRequest.newBuilder().setId(dbPullResponse.id).build())
+
+        Thread.sleep(1)
```

Comment:
thought: Why do you need a delay of 1 ms here? You probably try to wait until the pulling task stops. But there is no guarantee that for some reason it might take longer than 1 ms... And the test might fail in some cases. Maybe it is better to stop the application completely and start it again. Or add an API to check if the task is finished - like a callback that will complete gRPC call once the task really finished (but I think the first option is the easiest one)

Reply:
I think the stop pulling method should grantee stop the task before return respond to a gRPC call side